File: sdks/ios/samples/VideoHeartbeatSample/VideoHeartbeatSample/analytics/VideoAnalyticsProvider.m
/*
 * ADOBE SYSTEMS INCORPORATED
 * Copyright 2014 Adobe Systems Incorporated
 * All Rights Reserved.

 * NOTICE:  Adobe permits you to use, modify, and distribute this file in accordance with the
 * terms of the Adobe license agreement accompanying it.  If you have received this file from a
 * source other than Adobe, then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

#import <MediaPlayer/MediaPlayer.h>
#import "VideoAnalyticsProvider.h"
#import "ADB_VHB_VideoHeartbeat.h"
#import "VideoPlayer.h"
#import "VideoPlayerDelegate.h"
#import "Configuration.h"

@interface VideoAnalyticsProvider ()

@property(nonatomic, assign) VideoPlayer *player;
@property(nonatomic, retain) ADB_VHB_VideoHeartbeat *videoHeartbeat;
@property(nonatomic, retain) VideoPlayerDelegate *playerDelegate;

@end

@implementation VideoAnalyticsProvider

#pragma mark - Allocation/de-allocation

- (id)init {
    self = [super init];
    if (self) {
        [NSException raise:@"Illegal invocation." format:@"Use the initWithPlayer: selector."];
    }

    return self;
}

- (id)initWithPlayer:(VideoPlayer *)player {
    self = [super init];

    if (self) {
        if (!player) {
            [NSException raise:@"Illegal argument." format:@"Player reference cannot be NULL."];
        }

        _playerDelegate = [[VideoPlayerDelegate alloc] initWithPlayer:player provider:self];
        _videoHeartbeat = [[ADB_VHB_VideoHeartbeat alloc] initWithPlayerDelegate:_playerDelegate];

        _player = player;

        [self setupVideoHeartbeat];
        [self setupPlayerNotifications];
    }

    return self;
}

- (void)dealloc {
    [self tearDown];

    [super dealloc];
}

#pragma mark - VideoPlayer notification handlers.

- (void)onMainVideoLoaded:(NSNotification *)notification {
    [self.videoHeartbeat trackVideoLoad];
}

- (void)onMainVideoUnloaded:(NSNotification *)notification {
    [self.videoHeartbeat trackVideoUnload];
}

- (void)onPlay:(NSNotification *)notification {
    [self.videoHeartbeat trackPlay];
}

- (void)onStop:(NSNotification *)notification {
    [self.videoHeartbeat trackPause];
}

- (void)onSeekStart:(NSNotification *)notification {
    [self.videoHeartbeat trackSeekStart];
}

- (void)onSeekComplete:(NSNotification *)notification {
    [self.videoHeartbeat trackSeekComplete];
}

- (void)onComplete:(NSNotification *)notification {
    [self.videoHeartbeat trackComplete];
}

- (void)onChapterStart:(NSNotification *)notification {
    [self.videoHeartbeat trackChapterStart];
}

- (void)onAdChapterComplete:(NSNotification *)notification {
    [self.videoHeartbeat trackChapterComplete];
}

- (void)onAdStart:(NSNotification *)notification {
    [self.videoHeartbeat trackAdStart];
}

- (void)onAdComplete:(NSNotification *)notification {
    [self.videoHeartbeat trackAdComplete];
}

#pragma mark - Private helper methods

- (void)setupVideoHeartbeat {
    ADB_VHB_ConfigData *videoHeartbeatConfig = [[[ADB_VHB_ConfigData alloc] initWithTrackingServer:TRACKING_SERVER
                                                                                             jobId:JOB_ID
                                                                                         publisher:PUBLISHER] autorelease];

    videoHeartbeatConfig.channel = @"test_channel";

    // Set this to true to activate the debug tracing.
    // NOTE: remove this in production code.
    videoHeartbeatConfig.debugLogging = YES;

    [_videoHeartbeat configure:videoHeartbeatConfig];
}

- (void)setupPlayerNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onMainVideoLoaded:)
                                                 name:PLAYER_EVENT_VIDEO_LOAD
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onMainVideoUnloaded:)
                                                 name:PLAYER_EVENT_VIDEO_UNLOAD
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onPlay:)
                                                 name:PLAYER_EVENT_PLAY
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onStop:)
                                                 name:PLAYER_EVENT_PAUSE
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onSeekStart:)
                                                 name:PLAYER_EVENT_SEEK_START
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onSeekComplete:)
                                                 name:PLAYER_EVENT_SEEK_COMPLETE
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onComplete:)
                                                 name:PLAYER_EVENT_COMPLETE
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onChapterStart:)
                                                 name:PLAYER_EVENT_CHAPTER_START
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onAdChapterComplete:)
                                                 name:PLAYER_EVENT_CHAPTER_COMPLETE
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onAdStart:)
                                                 name:PLAYER_EVENT_AD_START
                                               object:self.player];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onAdComplete:)
                                                 name:PLAYER_EVENT_AD_COMPLETE
                                               object:self.player];
}

- (void)tearDown {
    // Detach from the notification center.
    [[NSNotificationCenter defaultCenter] removeObserver:self];

    [_videoHeartbeat release];
    _videoHeartbeat = nil;
    [_playerDelegate release];
    _playerDelegate = nil;
}

@end

File: sdks/js/samples/VideoHeartbeatSample/script/app/video.analytics.provider.js
(function(ADB, Configuration, VideoPlayerDelegate) {
    'use strict';

    var VideoHeartbeat = ADB.va.VideoHeartbeat;
    var ConfigData = ADB.va.ConfigData;

    function VideoAnalyticsProvider(appMeasurement, player) {
        if (!appMeasurement) {
            throw new Error("Illegal argument. AppMeasurement reference cannot be null.")
        }

        if (!player) {
            throw new Error("Illegal argument. Player reference cannot be null.")
        }
        this._player = player;

        this._playerDelegate  = new VideoPlayerDelegate(this._player, this);
        this._videoHeartbeat = new VideoHeartbeat(appMeasurement, this._playerDelegate);

        this._setupVideoHeartbeat();
        this._installEventListeners();
    }

    VideoAnalyticsProvider.prototype.destroy = function() {
        if (this._player) {
            this._videoHeartbeat.destroy();
            this._videoHeartbeat = null;
            this._playerDelegate = null;

            DefaultCommCenter().notificationCenter.removeAllListeners(this);
            this._player = null;
        }
    };


    /////////
    // Notification handlers
    /////////

    VideoAnalyticsProvider.prototype._onLoad = function(e) {
        console.log('Player event: VIDEO_LOAD');
        this._videoHeartbeat.trackVideoLoad();
    };

    VideoAnalyticsProvider.prototype._onUnload = function(e) {
        console.log('Player event: VIDEO_UNLOAD');
        this._videoHeartbeat.trackVideoUnload();
    };

    VideoAnalyticsProvider.prototype._onPlay = function(e) {
        console.log('Player event: PLAY');
        this._videoHeartbeat.trackPlay();
    };

    VideoAnalyticsProvider.prototype._onPause = function(e) {
        console.log('Player event: PAUSE');
        this._videoHeartbeat.trackPause();
    };

    VideoAnalyticsProvider.prototype._onSeekStart = function(e) {
        console.log('Player event: SEEK_START');
        this._videoHeartbeat.trackSeekStart();
    };

    VideoAnalyticsProvider.prototype._onSeekComplete = function(e) {
        console.log('Player event: SEEK_COMPLETE');
        this._videoHeartbeat.trackSeekComplete();
    };

    VideoAnalyticsProvider.prototype._onBufferStart = function(e) {
        console.log('Player event: BUFFER_START');
        this._videoHeartbeat.trackBufferStart();
    };

    VideoAnalyticsProvider.prototype._onBufferComplete = function(e) {
        console.log('Player event: BUFFER_COMPLETE');
        this._videoHeartbeat.trackBufferComplete();
    };

    VideoAnalyticsProvider.prototype._onAdStart = function(e) {
        console.log('Player event: AD_START');
        this._videoHeartbeat.trackAdStart();
    };

    VideoAnalyticsProvider.prototype._onAdComplete = function(e) {
        console.log('Player event: AD_COMPLETE');
        this._videoHeartbeat.trackAdComplete();
    };

    VideoAnalyticsProvider.prototype._onChapterStart = function(e) {
        console.log('Player event: CHAPTER_START');
        this._videoHeartbeat.trackChapterStart();
    };

    VideoAnalyticsProvider.prototype._onChapterComplete = function(e) {
        console.log('Player event: CHAPTER_COMPLETE');
        this._videoHeartbeat.trackChapterComplete();
    };

    VideoAnalyticsProvider.prototype._onComplete = function(e) {
        console.log('Player event: COMPLETE');
        this._videoHeartbeat.trackComplete();
    };


    /////////
    // Private helper functions
    /////////

    VideoAnalyticsProvider.prototype._setupVideoHeartbeat = function() {
        var configData = new ConfigData(Configuration.HEARTBEAT.TRACKING_SERVER,
            Configuration.HEARTBEAT.JOB_ID,
            Configuration.HEARTBEAT.PUBLISHER);

        configData.ovp = Configuration.HEARTBEAT.OVP;
        configData.sdk = Configuration.HEARTBEAT.SDK;
        configData.channel = Configuration.HEARTBEAT.CHANNEL;

        // Comment or explicitly set this to false for production sites.
        configData.debugLogging = true;


        this._videoHeartbeat.configure(configData);
    };

    VideoAnalyticsProvider.prototype._installEventListeners = function() {
        // We register as observers to various VideoPlayer events.
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.VIDEO_LOAD, this._onLoad, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.VIDEO_UNLOAD, this._onUnload, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.PLAY, this._onPlay, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.PAUSE, this._onPause, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.SEEK_START, this._onSeekStart, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.SEEK_COMPLETE, this._onSeekComplete, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.BUFFER_START, this._onBufferStart, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.BUFFER_COMPLETE, this._onBufferComplete, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.AD_START, this._onAdStart, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.AD_COMPLETE, this._onAdComplete, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.CHAPTER_START, this._onChapterStart, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.CHAPTER_COMPLETE, this._onChapterComplete, this);
        DefaultCommCenter().notificationCenter.addEventListener(PlayerEvent.COMPLETE, this._onComplete, this);
    };


    // Export symbols.
    window.VideoAnalyticsProvider = VideoAnalyticsProvider;
})(window.ADB, window.Configuration, window.VideoPlayerDelegate);
